python/ray/data/_internal/arrow_block.py

```diff
@@ -204,21 +205,10 @@ def to_numpy(
                     f"Cannot find column {column}, available columns: "
                     f"{self._table.column_names}"
                 )
-        arrays = []
-        for column in columns:
-            array = self._table[column]
-            if array.num_chunks == 0:
-                array = pyarrow.array([], type=array.type)
-            elif _is_column_extension_type(array):
-                array = _concatenate_extension_column(array)
-            else:
-                array = array.combine_chunks()
-            arrays.append(array.to_numpy(zero_copy_only=False))
-        if len(arrays) == 1:
-            arrays = arrays[0]
-        else:
-            arrays = dict(zip(columns, arrays))
-        return arrays
+
+        numpy_batch = _convert_batch_type_to_numpy(self._table)
+        numpy_batch = dict(zip(columns, numpy_batch.values()))
+        return numpy_batch
```

Comment:
So this will break the current Arrow --> NumPy conversion semantics when `columns = None`, which is currently:
- Return a single NumPy ndarray for a single-column table.
- Return a dict of NumPy ndarray for a multi-column table.

This PR would change it to always return a dict of NumPy ndarray, even when the column name is our internal `TENSOR_COL_NAME`, which should definitely be returned as a single NumPy ndarray.

IMO we should align to the `_convert_batch_type_to_numpy()` semantics when `columns: List[str]` and `columns = None`, which are:
- Return a single NumPy ndarray for a single-column table with column name `TENSOR_COL_NAME`.
- Return a dict of NumPy ndarray otherwise.

This PR also currently changes the `columns: str` case, which should always return a single NumPy ndarray but with this PR would return a dict of ndarrays.

Overall, I think that the semantics should be:
- `columns = None or columns = List[str]`: `_convert_batch_type_to_numpy()` semantics.
- `columns = str`: Always return a single NumPy ndarray.

```python
        if columns is None:
            columns = self._table.column_names
        # Should add this helper to ArrowBlock.
        self._check_has_columns(columns)
        if not isinstance(columns, list):
            batch = self._table.column(columns)
        else:
            batch = self._table.select(columns)
        return _convert_batch_type_to_numpy(batch)
```

Also, we should make this same change in semantics + use of `_convert_batch_type_to_numpy()` in [`PandasBlock.to_numpy()`](https://github.com/ray-project/ray/blob/d527ce8b7d6367f83292e5da02fc3ebe6f11860a/python/ray/data/_internal/pandas_block.py#L193-L213).